src/Components/Components/src/Virtualization/VirtualizeBase.cs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Components.Rendering;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides common functionality for virtualized lists.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public abstract class VirtualizeBase<TItem> : ComponentBase, IAsyncDisposable
+    {
+        private int _itemsAbove;
+
+        private int _itemsVisible;
+
+        private ElementReference _topSpacer;
+
+        private ElementReference _bottomSpacer;
+
+        private IVirtualizationHelper? _virtualizationHelper;
+
+        [Inject]
+        private IVirtualizationService VirtualizationService { get; set; } = default!;
+
+        /// <summary>
+        /// Gets the total number of items in the collection.
+        /// </summary>
+        protected abstract int ItemCount { get; }
+
+        /// <summary>
+        /// Gets the size (height) of each item in pixels.
+        /// </summary>
+        [Parameter]
+        public float ItemSize { get; set; }
+
+        /// <summary>
+        /// Gets the <see cref="RenderFragment"/>s representing a range of items.
+        /// </summary>
+        /// <param name="start">The start index of the range of items.</param>
+        /// <param name="count">The number of items in the range.</param>
+        /// <returns>The <see cref="RenderFragment"/>s representing the given range.</returns>
+        protected abstract IEnumerable<RenderFragment> GetItems(int start, int count);
+
+        /// <inheritdoc />
+        protected override void OnParametersSet()
+        {
+            if (ItemSize <= 0f)
+            {
+                throw new InvalidOperationException($"Parameter '{nameof(ItemSize)}' must be specified and greater than zero.");
+            }
+        }
+
+        /// <inheritdoc />
+        protected override async Task OnAfterRenderAsync(bool firstRender)
+        {
+            if (firstRender)
+            {
+                _virtualizationHelper = VirtualizationService.CreateVirtualizationHelper();
+                _virtualizationHelper.TopSpacerVisible += OnTopSpacerVisible;
+                _virtualizationHelper.BottomSpacerVisible += OnBottomSpacerVisible;
+
+                await _virtualizationHelper.InitAsync(_topSpacer, _bottomSpacer);
+            }
+        }
+
+        /// <inheritdoc />
+        protected override void BuildRenderTree(RenderTreeBuilder builder)
+        {
+            var itemsBelow = Math.Max(0, ItemCount - _itemsVisible - _itemsAbove);
+
+            builder.OpenElement(0, "div");
+            builder.AddAttribute(1, "key", "top-spacer");
```

Comment:
It's possible I'm wrong, but I *think* that the `key` attributes won't be necessary here. Your use of `OpenRegion` on line 82 means that nothing within it can interfere with what's outside it for the purposes of diffing, so we can be sure of matching up and preserving the spacers anyway.

If I'm right about that, it's marginally preferable not to use `key` because then the diffing system doesn't have to build a dictionary of keys on each diff. Not a big deal at all, just a very minor suggestion!